Feature: permit running anything-sync-daemon as an unprivileged user

tomeon opened the pull request:
Plus a bunch of other fixes and features.  Note that this changeset includes major refactorings; I offer this PR with the expectation that its scope is too large to effectively review and therefore not reasonable to merge.  However, I've done my best to test and document the changes, and I think that they are genuinely useful, so if you've got the inclination to review this -- many thanks in advance!

Summary of changes:

1. Permit running `asd` as an unprivileged user.  This includes installing a systemd user unit, and loading `asd.conf` from a different location than `/etc/asd.conf`.  I've also added the script `asd-mount-helper`, based on [`psd-overlay-helper`](https://github.com/graysky2/profile-sync-daemon/blob/master/common/profile-sync-daemon.in), to facilitate mounting and unmounting as a non-`root` user.
2. Add an `anything-sync-daemon` Nix package (`nix/packages.nix`).
3. Add a NixOS module (`nix/nixos-module.nix`) for configuring and running system and user instances of `asd.service`, `asd-resync.service`, and `asd-resync.timer`.
4. Add NixOS-based systems tests (`nix/checks.nix`) that verify the correct operation of system and user `asd` services.
5. Add Nix-based development tooling (`nix/devshell.nix`), including the `shellcheck` shell linter and `alejandra` Nix expression language formatter.
6. Update the GitHub actions configuration to run the NixOS-based systems tests (and replace the `shellcheck` workflow, as an equivalent job is included in the Nix checks).
7. Translate `doc/asd.1` to Markdown (in `USAGE.md`) and generate the manpage from that Markdown.  Also expand the documentation to cover environment variables, and add some relevant links.
8. Factor out sync target iteration (repeated definitions of `DIR`, `BACKUP`, `TMP`, and so on).

I'm still ironing out the GitHub Actions workflow; currently it is quite slow (at least, the NixOS systems tests are).  But you can see a recent run [here](https://github.com/tomeon/anything-sync-daemon/actions/runs/6355418819).

Thanks again for your consideration!

tomeon committed:
feat: add Nix integration

along with a `numtide/devshell`-based development shell providing a
command to lint and format shell and Nix code (with, respectively,
shellcheck and alejandra).

tomeon committed:
feat(nix): add Nix derivation

and expand the `mkFlake` `systems` argument to more supported Linux
architectures.

tomeon committed:
feat(nix): add `services.asd` NixOS module

for managing systemwide and per-user anything-sync-daemon installations.

tomeon committed:
feat(nix): add NixOS test

that exercises various anything-sync-daemon features and behaviors,
including OverlayFS support and crash-recovery backups.

tomeon committed:
fix: use portable shebang

tomeon committed:
feat: add fancy PS4

that prints additional context like the source file and current line
number when Bash execution tracing is in effect.

tomeon committed:
fix: don't recheck `DEBUG` upon each `debug` call

but instead define the `debug` function so that it is a NOP when `DEBUG`
is disabled.

tomeon committed:
refactor: add truthiness-testing function

rather than inlining nearly-identical `case` statements for both
and `USE_OVERLAYFS` and `USE_BACKUPS`.

tomeon committed:
feat: add logging functions

`diag` (simple echo to stderr) and `ediag` (`echo -e` to stderr).  The
intention is to replace uses of `echo` and `echo -e` that write to
stdout.

tomeon committed:
feat: add (?:un)?mounting helper script

in preparation for supporting running anything-sync-daemon as a non-root
user.

tomeon committed:
feat: enable "strict mode"

that is, `set -euo pipefail`.

tomeon committed:
chore: issue message upon ERR but do not bail out

as we'll be bailing out anyway due to `set -e`.

Quote `help trap`:

    A SIGNAL_SPEC of ERR means to execute ARG each time a command's
    failure would cause the shell to exit when the -e option is enabled.

tomeon committed:
chore: consolidate dependency checks

with the `have_dep` function and its verbose wrapper `have_dep_diag`.

This de-duplicates code in `dep_check`, and also permits code re-use
when checking whether `flock` is available toward the beginning of the
script.

Additionally, clean up the OverlayFS version check in `dep_check` by (a)
using the arithmetic comparison operator `(( ... ))` and (b)
consolidating to a single conditional statement.

Finally, don't bail out at the first failed dependency, but instead
store the non-zero exit status and continue.  This way, the code will
report all missing dependencies in one go.

tomeon committed:
feat: improve (f)locking logic

by (a) handling `FLOCKER` being unset, (b) respecting the choice of Bash
executable by prepending `"$BASH"` to the re-executed command line, (c)
propagating the current shell options by passing `"-$-"`, and (d) using
portable `flock` options (omit `--verbose`, use `-x` rather than `-e`).

tomeon committed:
feat: remove asd.conf syntax check

and replace it with explicit error messaging mentioning the
configuration file.

tomeon committed:
chore: s/\./source/

tomeon committed:
feat: make shellcheck find asd.conf next to asd

via the `source-path=SCRIPTDIR` directive.

tomeon committed:
chore: simplify variable assignments

using the `:=` "default-assignment" parameter expansion operator.

tomeon committed:
chore: remove unneeded `""` in `debug ""`

tomeon committed:
chore: s/true/:/

tomeon committed:
fix: handle unset WHATTOSYNC array

tomeon committed:
fix: suppress errors re-enforcing extglob setting

tomeon committed:
fix: handle unset `VOLATILE` variable

tomeon committed:
chore: put "VOLATILE is tmpfs/zram" check in func

rather than inlining it.

tomeon committed:
chore: simplify "can I use OverlayFS?" check

to (a) replace `grep -c` with `grep -q` (we only care if there was a
match, not how many matches there are -- incidentally this fixes a bug
where > 1 match would make `asd` liable to set `OLFSVER` to the wrong
value, or leave it unset) and (b) restructure the checks as a single
level of conditionals, rather than nesting conditionals.

tomeon committed:
feat: more portable "using systemd?" check

that uses `systemctl list-unit-files` in preference to checking for a
unit file at `/usr/lib/systemd/system/asd.service`.

tomeon committed:
chore: simplify distro check and header definition

tomeon committed:
feat: add function for running cmds verbosely

thus laying the groundwork for simplifying code like:

    debug "About to do a thing..."
    if do-thing; then
      debug "Did the thing!"
    else
      echo "Oh dear, failed to do the thing :(" >&2
    fi

tomeon committed:
feat: add a for-each-WHATTOSYNC-dir function

that encapsulates the logic involved in performing operations on each
entry in the `WHATTOSYNC` array.

This lays the groundwork for de-duplicating `WHATTOSYNC` iteration code.

tomeon committed:
feat: add "running as root?" function

in preparation for allowing anything-sync-daemon to run as a non-root
user.

tomeon committed:
feat: add "do I need/have sudo?" functions

in preparation for permitting anything-sync-daemon to run as a non-root
user.

tomeon committed:
feat: permit running asd as non-root user

This is a major refactor with a number of incidental changes.

Summary of changes:

  1. Replace inline `WHATTOSYNC` iteration with the use of
     `for_each_dir` plus appropriate callback functions.
  2. Use `asd-mount-helper` for mounting and unmounting.
  3. Replace inline variable-dumping code with `debug_vars` function
     calls.
  4. Detect systemd user instance `anything-sync-daemon.service` unit
     file availability, and
  5. Use systemd's `CONFIGURATION_DIRECTORY` and `RUNTIME_DIRECTORY`
     variables, if defined, falling back to appropriate defaults (e.g.
     `${XDG_CONFIG_HOME}/asd` for non-root users).

tomeon committed:
feat: respect the `NO_COLOR` env var

See https://no-color.org/.

tomeon committed:
chore: miscellaneous linting and formatting

tomeon committed:
fix(init/asd.service): persist RuntimeDirectory

so that the daemon file outlasts the oneshot service.  This means that
`asd-resync.service` will correctly detect that `asd.service` ran (if
indeed it did run).

tomeon committed:
feat: user-settable `ASD{CONF,RUN}DIR` env vars

tomeon committed:
fix: restore `ASDCONF` and `VOLATILE` defaults

when running as root.  The user may override this by setting the
`ASDNOV1PATHS` environment variable.

tomeon committed:
fix(nixos): set `ASDNOV1PATHS=yes`

in the generated systemd service definition.

tomeon committed:
fix: stop asd.conf from redefining important vars

by marking them as readonly with the eponymous builtin.

BREAKING CHANGE: some configuration files may redefine these variables.
This change is liable to make such code bail out; failing that, whatever
differences such code would have wrought are now null and void.

tomeon committed:
chore: ignore `.nixos-test-history`

tomeon committed:
docs(asd.conf): note that file is a Bash script

and that precautions are appropriate.

tomeon committed:
feat(nix): add user unit NixOS test cases

tomeon committed:
feat: enforce timeout sourcing `asd.conf`

to help catch and report pathological conditions (e.g. starting a
long-running program).

BREAKING CHANGE: because `asd.conf` is now sourced in a subprocess, it
cannot *directly* affect the execution state of `anything-sync-daemon`
via (for instance) (re)defining functions, or setting shell options, or
(re)defining any variables other than those in the list of variables
that are supposed to be set in `asd.conf` (`WHATTOSYNC`, `VOLATILE`, and
so on).  It's possible that this change will therefore break `asd.conf`
setups in the wild, though likely only setups that were doing something
arguably naughty.

BREAKING CHANGE: enforcing a timeout on sourcing `asd.conf` may break
setups that (for whatever reason) genuinely require one or more
long-running operations.  This can be addressed by setting the
newly-introduced `ASDCONFTIMEOUT` environment variable to a value even
laxer than the default `10` (seconds).

tomeon committed:
feat(ci): add NixOS systems tests workflow

tomeon committed:
docs: add USAGE.md

which is a Markdown translation of ./doc/asd.1.

tomeon committed:
feat: build asd manpage from USAGE.md with pandoc

tomeon committed:
chore(asd): s/Overlayfs/OverlayFS/g

tomeon committed:
chore: regenerate manpage

tomeon committed:
docs(USAGE): rework structure and add links

and use other Markdown features and Pandoc extensions to promote
usability as a manpage and as a webpage hosted on GitHub.

tomeon committed:
docs(README): rework structure and add links

including a link to the new USAGE.md document.

tomeon committed:
fix: check mountpoints in ungraceful state check

to avoid performing `rsync` only in `do_sync` without first ensuring
that all bind mounts (and, if appropriate, OverlayFS mounts) are in
place.

tomeon committed:
chore(asd): move negation outside `[[ ... ]]`

tomeon committed:
feat(checks): add ungraceful state subtests

tomeon committed:
feat(nix): rework test node into example profile

tomeon committed:
feat: add `nixos-shell` for running a dev VM

tomeon committed:
refactor(asd): add `main` function

that processes options and dispatches to the appropriate operation
(parse, sync, etc.).

tomeon committed:
fix(nix): rework crash-recovery subtest

to eschew crashing the entire machine in favor of simply removing the
`.flagged` files from mounted tmpfs dirs.  This method is more reliable
at triggering crash-recovery file generation, and is also faster.

tomeon committed:
refactor: move setup steps into functions

and out of the script toplevel.

tomeon committed:
feat: obtain lock in-process

rather than re-executing if the `FLOCKER` environment variable is unset
or empty.

tomeon committed:
fix: check that `WHATTOSYNC` is declared

in `config_check`, in addition to checking its size.

tomeon committed:
feat: support sourcing `anything-sync-daemon`

tomeon committed:
chore: use printf builtin for crash file dates

falling back to the `date` command if using the `%(...)T` printf format
fails.

tomeon committed:
chore(ci): remove `shellcheck` workflow

as the Nix flake checks include running `shellcheck`.

tomeon committed:
refactor: add `run_systemd` function

that abstracts over `systemctl --system` and `systemctl --user`,
depending on the EUID that's running the script.

tomeon committed:
feat: detect systemd units besides `asd.service`

tomeon committed:
feat: display systemd unit name

in debugging output, if available.

tomeon committed:
refactor: use internal-only settings

instead of directly consulting variables (potentially) defined in
`asd.conf`.  This makes `asd` more resilient in the face of things like
this:

    # asd.conf
    declare -gar WHATTOSYNC=(/haha/good/luck// /modifying/this///)

Two approaches are used:

    1. Declaring functions for boolean conditions.  For instance, the new
       `use_overlayfs` returns status 0 if `USE_OVERLAYFS` is set to a
       truthy value, and returns `1` otherwise.
    2. Declaring internal variables.  These follow the naming convention
       "<user-facing-variable-name>_FINAL"; for instance, instead of
       using `VOLATILE`, `asd` now uses `VOLATILE_FINAL`.

This changeset also renames certain variables to make them consistent
with other variable names (e.g. renaming `OLFSVER` to `OVERLAYFS_VERS`,
making it consistent with `USE_OVERLAYFS` and `VERS`).

tomeon committed:
fix: respect `DEBUG` definition from `asd.conf`

by re-declaring the `debug` function after sourcing the `asd.conf`
snapshot.

tomeon committed:
refactor: add `for_each_dir_with_crasharr` func

that wraps `for_each_dir`, with `NEED_CRASHArr` set to a truthy value.

tomeon committed:
chore: regenerate man page

tomeon committed:
docs: describe significant environment variables

tomeon committed:
feat(devshell): add gnumake and gzip

to available packages, enabling things like running `make install-man`
from within the devshell.

tomeon committed:
docs: add BleachBit link

tomeon committed:
docs: describe `DEBUG` variable

tomeon committed:
chore: git-ignore generated script

tomeon committed:
feat(Makefile): install systemd user unit files

tomeon committed:
fix(Makefile): install conf directory

tomeon committed:
fix(Makefile): overwrite compressed manpage

and prevent being interactively prompted.

tomeon committed:
docs(INSTALL): flesh out prereqs list

tomeon committed:
fix(INSTALL): correct typos and out-of-date info

tomeon committed:
feat(nix): add `docs` package

that provides Markdown-based documentation of
`nixosModules.anything-sync-daemon`.

tomeon committed:
feat(nix): add devshell command to generate docs

tomeon committed:
feat(nix): add "docs up-to-date?" check

tomeon committed:
docs: add NixOS module docs

tomeon committed:
fix(nix): describe `configFile` option

tomeon committed:
docs(INSTALL): describe Nix package, NixOS module

tomeon committed:
feat(ci): parallelize jobs

and cache the Nix store.

tomeon committed:
feat(devshell): install act

which is a local GitHub Actions runner (https://github.com/nektos/act).

tomeon committed:
fix(ci): empty Nix `build-users-group`

to permit running `nix-quick-install-actions` under `act`.

tomeon committed:
fix(ci): use local flake.nix for building nix conf

tomeon committed:
fix(ci): don't cache Nix store in local runs

tomeon committed:
feat(ci): build NixOS test driver for debugging

when running locally.

tomeon committed:
fix(nix): don't wait for `multi-user.target`

in the `anything-sync-daemon` Nix flake check, as it (a) appears to be
unnecessary to do so (even though we need an `asduser` session to run
certain assertions) and (b) appears to break the check on GitHub
actions, as we hit the retry timeout before `multi-user.target` is
reached.

tomeon committed:
feat: generic path-cleaning implementation

that removes trailing slashes and (if either `realpath -m` or `readlink
-m` are available) also cleans up interior duplicated slashes, useless
".", etc.

tomeon committed:
feat: support shell expansions in whatToSync opt

by removing the use of `lib.escapeShellArgs` in the
configuration-file-generation step.

tomeon committed:
feat: deduplicate sync targets

and warn if a path was specified more than once in `WHATTOSYNC`.

tomeon committed:
feat: add debugging output re mount state

to aid in diagnosing issues with asd's mounts

tomeon committed:
feat: issue warnings upon unexpected mounts

tomeon committed:
feat: add yellow color

tomeon committed:
fix: explicitly pass `--list` to `findmnt`

to avoid getting prettified output.

tomeon committed:
fix: find only *first* matching mount

to avoid getting back data for multiple mounts when we just want one.

tomeon committed:
fix: reliably identify canonical bind mount source

tomeon committed:
chore: move `verbosely` to `run_mount_helper`

tomeon committed:
chore: shell-quote commands in `verbosely`

to better delineate what's being run

tomeon committed:
chore: use `verbosely` instead of `set -x`

to provide additional diagnostics (e.g. exit status) and to avoid
clobbering `xtrace` if it's already enabled.

tomeon committed:
fix: remove symlink-listing tempfile on error

tomeon committed:
feat: use NUL separator between symlink names

when caching them for later synchronization, thus better handling funky
filenames.

tomeon committed:
fix(checks): block until flag file exists

before shutting off the resync service.

tomeon committed:
chore(ci): split checks into "system" and "user"

in an attempt to speed up GitHub Actions by running the checks
concurrently.  Involves some not-terribly-attractive monkeypatching,
with attendant stubbing/casting/etc. to satisfy MyPy.

manorit2001 commented:
Hi @tomeon, I really like the changes that you have been doing ( just a bird's eye review based on your commits ). Unfortunately , I haven't been able to spend much time to give a detailed review as you already know the changes are humongous and it's not exactly possible to review but I do see that you are active and doing some good improvements based on your commits. I will see if I can review few of the stuff but I can't commit at this stage.

Although, I would like to know a few things. Is there any breaking change in this full PR w.r.t the current asd version and something that you would like to summarise based on that? The debug capability handling along with fixes and added nix os support is always welcomed but the only thing that I am mostly worried about is the compatibility with the current asd version. I see that asd is a very critical piece of software and people have actually suffered data losses in the past due to the refactor and the syncing process, if you know anything that you have done to drastically alter that process ( along with changing any names for the backup folders or anything), I would recommend giving a summary of it and maybe explaining the changes here ( or if possible pointing to those specific commits along with the explanation ) as that'd help us see the scope of the upgradation process from current asd to the new one. Also, I see that most of your testing has been NixOS based, is there anything that you have done for arch or is there some difference in that? I haven't really ventured into NixOS so feel free to impart your knowledge

tomeon committed:
fix(asd.service): comment out `RuntimeDirectory`

and explain that it should not be used unless `ASDNOV1PATHS` is enabled.

tomeon committed:
feat: add Vagrant configuration

for spawning an Arch Linux machine and provisioning `asd`.

tomeon commented:
Hello @manorit2001 -- thank you very much for your feedback. 

> Is there any breaking change in this full PR w.r.t the current asd version and something that you would like to summarise based on that?

If you subscribe to [Hyrum's Law](https://www.hyrumslaw.com/), then I'm afraid there are breaking changes in just about every commit :wink: .  More seriously: a few of my changes are liable to break some setups, particularly the changes having to do with loading `asd.conf` and its snapshot.  However: setups broken by these changes are probably doing something (a) naughty (like messing with `anything-sync-daemon`'s internal state) or (b) inadvisable (like launching long-running processes).  In case (a), users shouldn't do that, and in case (b) there's the workaround/escape hatch of specifying a longer timeout in `ASDCONFTIMEOUT`.

> I see that asd is a very critical piece of software and people have actually suffered data losses in the past due to the refactor and the syncing process, if you know anything that you have done to drastically alter that process ( along with changing any names for the backup folders or anything), I would recommend giving a summary of it and maybe explaining the changes here ( or if possible pointing to those specific commits along with the explanation ) as that'd help us see the scope of the upgradation process from current asd to the new one.

I am working on a `CHANGELOG.md` that will cover all significant changes, highlighting especially any breaking changes.  I have not changed the names of backup folders, and any path-related changes are strictly opt-in, either by (a) running `anything-sync-daemon` as an unprivileged user (so should not affect existing installations, though I cannot say this for certain) or (b) setting `ASDNOV1PATHS=1` or similar, which is not the default.

> I see that most of your testing has been NixOS based, is there anything that you have done for arch or is there some difference in that? I haven't really ventured into NixOS so feel free to impart your knowledge

I hadn't done anything for Arch Linux, though, inspired by your mention of it, I've added a Vagrant configuration that spins up an Arch Linux machine and deploys `anything-sync-daemon`.  Found a few bugs in my other changes as a result, so thanks for that :slightly_smiling_face:.  

Part of what I like about NixOS systems tests is how the testing framework is distributed with `nixpkgs`, so if you've got the Nix package manager and an internet connection, you're only a few steps from rockin' and rollin' with `pkgs.testers.nixosTest`.  No need to explicitly install additional tooling, or even configure much of anything outside the actual stuff of interest (in this case, settings in the `config.services.asd` "namespace"), as `nixpkgs` handles all the QEMU/KVM/networking/etc. stuff out-of-the-box.

However, I am happy to look into rigorous testing on Arch Linux, too.  Haven't found any analog of `pkgs.testers.nixosTest` for Arch, though perhaps [Serverspec](https://serverspec.org) or [InSpec](https://docs.chef.io/inspec) or similar would do the trick.  WDYT?

tomeon committed:
fix: handle > 1 {CONFIGURATION,RUNTIME}_DIRECTORY

by selecting the first of the colon-separated paths.  This is necessary,
as systemd supports multiple directories specified as
`ConfigurationDirectory=` and `RuntimeDirectory=`.

tomeon committed:
docs: add changelog

[skip ci]

tomeon commented:
I have added the changelog.

tomeon commented:
Also, latest run is passing; see [here](https://github.com/tomeon/anything-sync-daemon/actions/runs/6450295604).

tomeon committed:
feat: warn about the use of V1 paths

tomeon committed:
docs: further usage updates and clarifications

[skip ci]

manorit2001 commented on Makefile:
```diff
@@ -15,6 +17,10 @@ BSHDIR = $(PREFIX)/share/bash-completion/completions
 # set to anything except 0 to enable manpage compression
 COMPRESS_MAN = 1
 
+# set to 1 to prevent attempting to stop asd before installation
+SKIP_STOP = 0
```
Although added but I am not sure when it'll be set to `1` actually, the way to upgrade asd should always disable asd before updating for good sanity upgrades as it can happen that the updated asd unsync might not be able to handle the changes and would break the sync that would've happened by previous asd version but I believe this can be a good thing for testing where there wouldn't be a lot of breaking changes but just wanted to let you know about the rational behind stopping asd before upgrades

tomeon replied:
I've used this for manual testing; will remove it.

manorit2001 commented on Makefile:
```diff
@@ -71,10 +86,14 @@ install-cron:
+	$(INSTALL_DATA) init/asd-resync.service "$(DESTDIR)$(INITDIR_SYSTEMD_SYSTEM)/asd-resync.service"
+	$(INSTALL_DATA) init/asd-resync.timer "$(DESTDIR)$(INITDIR_SYSTEMD_SYSTEM)/asd-resync.timer"
+	$(INSTALL_DIR) "$(DESTDIR)$(INITDIR_SYSTEMD_USER)"
+	$(INSTALL_DATA) init/asd.service "$(DESTDIR)$(INITDIR_SYSTEMD_USER)/asd.service"
```
don't both the system level and user level asd.server conflict with each other?

tomeon replied:
I don't think so; at least, not without the user somehow configuring them to conflict with one another.  Under their default configurations, the system daemon and user-specific daemons:

1. Use different systemd unit files,
2. Use different configuration files,
3. Use different daemon files,
4. Use different lock files,
5. Run as different users,
6. Etcetera.

manorit2001 replied:
ah okay, if that is the case then should be okay, wanted to know if these created any troubles or not in your manual testing. If no visible troubles then should be okay.

manorit2001 commented on README.md:
```diff
@@ -1,11 +1,20 @@
+
+Anything-sync-daemon (asd) is a tiny pseudo-daemon designed to manage user
+defined dirs in tmpfs and to periodically sync back to the physical disc
+(HDD/SSD). This is accomplished via a symlinking step and an innovative use of
```
not symlinking step anymore, I believe it has been changed to bind mounts but a stale documentation, maybe you can update that as well. Can see the arch wiki to be updated as well as your description of `Usage.md` so maybe fix the stale documents also  from there.

tomeon replied:
Sure; happy to handle these items.

manorit2001 commented on README.md:
```diff
@@ -1,11 +1,20 @@
-Consult the man page or the wiki page: https://wiki.archlinux.org/index.php/Anything-sync-daemon
+
+Consult the `asd(1)` man page (available [here](/USAGE.md)) or
+[the Arch Linux `asd` wiki page](https://wiki.archlinux.org/index.php/Anything-sync-daemon).
```
If you've added a nix package or something maybe provide that path also here

tomeon replied:
There is a Nix package, but it is distributed as a part of this repository and there is no external documentation specific to the NixOS package.  I could link to the package definition itself?  That is, link to `nix/packages.nix`?

manorit2001 replied:
ah okay, you can skip it that then, I was wondering if this gives information about what all linux variants are supported and in first glance of the readme, it doesn't look like nix is supported, maybe we can explicitly mention it then.

manorit2001 commented on USAGE.md:
```diff
@@ -0,0 +1,383 @@
+
+# NAME
+
+`anything-sync-daemon` - Symlinks and syncs user specified dirs to RAM thus
```
again symlinking here as well.

tomeon replied:
:+1:

manorit2001 commented on USAGE.md:
```diff
@@ -0,0 +1,383 @@
+
+`ASDNOV1PATHS`
+
+: A boolean variable controlling whether to use "version 1" paths or not.  When
+  set to a false value, `asd` will use old-style defaults for its daemon file
+  (`/run/asd`), lock file (`/run/asd-lock`), and configuration file
+  (`/etc/asd.conf`).  **NOTE** that this variable is ignored when `asd` is
+  running as a non-`root` user.  Defaults to disabled (effectively,
+  `ASDNOV1PATHS=0`).
```
What are the paths when this is not enabled? Also, I think renaming this to `ASD_BACKWARD_COMPAT` would be good ig as it seems like it is something similar to that, this name is a bit confusing for me.

tomeon replied:
> What are the paths when this is not enabled?

I'll make sure to enumerate these in the documentation.  For the systemd `asd.service` unit, the locations are as follows:
1. Daemon file, in order of precedence from highest to lowest: 
    a. the value of the `DAEMON_FILE` environment variable, if set and non-empty,
    b.  `${ASDRUNDIR}/asd.run`, if `ASDRUNDIR` is set and non-empty,
    c.  `${RUNTIME_DIRECTORY%%:*}/asd.run`, if `RUNTIME_DIRECTORY` is set and non-empty (note that `RUNTIME_DIRECTORY` is a list of colon-separated directories set by systemd on the basis of the `RuntimeDirectory` directive from the `[Service]` section of a systemd unit), and, finally,
    d. `/run/asd/asd.conf`.
2. Lock file: pretty much the same as the daemon file, with these modifications: the basename is `asd.lock`, and there is no equivalent `DAEMON_FILE`-style environment variable that the user can set to an arbitrary path.
3. Configuration file: `/etc/asd/asd.conf`.

> I think renaming this to ASD_BACKWARD_COMPAT would be good ig as it seems like it is something similar to that, this name is a bit confusing for me.

I think that `ASD_BACKWARD_COMPAT` is too general; there are any number of ways in which `anything-sync-daemon` might become backward-incompatible in the future, which implies on or more of the following:

1. The meaning of `ASD_BACKWARD_COMPAT` would change over time (the behaviors it controls would differ),
2. Users would face an all-or-nothing decision: all backward-compatible behaviors, or none, or
3. We'd have to choose new variable names to control new backward-compatibility-or-not behavior.

I chose `ASDNOV1PATHS` for the following reasons:

1. Preexisting `ASD*` variables all eschew underscores and are all quite terse (though, for what it's worth, in general I like separating semantically-distinct elements with underscores),
3. The `NO` bit helps emphasize that this is opt-in behavior and that by default you get the old path scheme, and is less liable to confuse after `asd` fully deprecates the old path scheme (`ASDNOV1PATHS=no` will cease to have any effect, whereas `ASDNOV1PATHS=yes` will, in a sense, remain in effect forever).
4. The `V1` bit scopes the effect to a particular path scheme (that is, the _first_ one), and
5. The `PATHS` bit scopes the effect to a particular realm of concern (that is, paths significant to `asd`).

However, I'd be happy to consider alternatives that satisfy at least some of the desiderata just listed.  What do you think about something like `ASD_DISABLE_V1_FILE_PATH_SCHEME_COMPAT`?

manorit2001 replied:
Could you clarify why was there a need for this change? Does this help adding support for running asd as an unprivileged user? Basically I want to understand if there was a need or if this a stylistic change only.

manorit2001 commented on USAGE.md:
```diff
@@ -0,0 +1,383 @@
+
+graysky (graysky AT archlinux DOT us)
+
+# MAINTAINER
```
You can add a contributors section and add your name as well.

tomeon replied:
:+1:

manorit2001 commented on common/anything-sync-daemon.in:
```diff
@@ -1,287 +1,520 @@
+  ' "$0" "${1?}"
+}
+
+use_v1_paths() {
```
I am not sure about why we are changing paths btw, what benefits do we get and what paths are we ending up changing?

manorit2001 reviewed (commented):
Hey @tomeon, apologies for the delay but you know it's a big PR so had to see what all changes are being done, I still can't cover it thoroughly but trying to gauge the changes being done in this, have some questions as well. Why is shellcheck ci removed btw?

manorit2001 commented:
> However, I am happy to look into rigorous testing on Arch Linux, too. Haven't found any analog of `pkgs.testers.nixosTest` for Arch, though perhaps [Serverspec](https://serverspec.org) or [InSpec](https://docs.chef.io/inspec) or similar would do the trick. WDYT?

There is one test that I am particularly worried about first of all with this PR, it is basically the upgrade process, if possible could you check if we use the current asd version, stop asd ( this will happen as the part of upgradation process ) and then update your asd version to the new one with same configs as it was previously, I want to know that nothing breaks as a part of that. 

W.r.t inspec or serverspec, I don't have any strong preference as I haven't worked on them, I would recommend you to choose whatever feels scalable to you for the project, would see the changes based on that.

tomeon committed:
chore(gitignore): ignore Vagrant state

tomeon committed:
refactor: lib for generators, options, and modules

in preparation for creating a Home Manager module that re-uses some code
from the NixOS module.

tomeon committed:
doc(treewide): s/symlink/bind-mount/

as asd now uses bind-mounts rather than symlinks for exposing the
tmpfs-es that it manages.

tomeon committed:
feat(nix): add Home Manager module

tomeon committed:
WIP on Nix installation instructions

tomeon committed:
docs: add myself as a contributor

tomeon committed:
fix(nix): add more `asd` prereq packages

tomeon committed:
docs: document home-manager module

tomeon committed:
chore: add `_file` annotations to modules
